datafusion/core/src/physical_plan/joins/nested_loop_join.rs

```diff
@@ -304,6 +280,14 @@ async fn load_left_specified_partition(
     Ok(merged_batch)
 }
 
+// BuildLeft means the left relation is the single patrition side.
+// For full join, both side are single partition, so it is BuildLeft and BuildRight, treat it as BuildLeft.
+pub fn left_is_build_side(join_type: JoinType) -> bool {
```

Comment:
if I‘m right, this part is the most important insight for your fix.
If the left can be the single partition, we will iter the right table/partition
If the right is the single partition, we will iter the left table/partition

Reply:
Yes, the iter way is start from outer-table now. I think:
* It is more suitable for the meaning of nested loop join.
* We can optimize the data type of join_indices(not do in this pr). (using `UInt64Array` for inner-table data, and using `UInt32Array` for the outer-table data)

Reply:
> * We can optimize the data type of join_indices(not do in this pr). The indices type of inner-table should be `UInt64Array`, and  `UInt32Array` for the outer-table. Currently left table is `UInt64Array`, and right table is `UInt32Array`.

Hash join has the same issue.

Before this PR, the NLJ and hash join share some methods in the `utils`, such as `need_produce_result_in_final`.
But this PR breaks this, I want to refactor them, Do you have any ideas about that?

In the hash-join, we always build left table and cache the left table(single partition or multi partition), and the right data will be visited iteratively with the `INT32 ARRAY` as the index.

Reply:
Current the indices type are fixed in the following functions, it is not enough, so we should make them more generic first.
```text
apply_join_filter_to_indices
build_batch_from_indices
adjust_indices_by_join_type
get_semi_indices
get_anti_indices
```

Then we can choose array corresponding the distribution(left and right), like:
* (Single, Unspecified) -> (UInt64Array, UInt32Array)
* (Unspecified, Single) -> (UInt32Array, UInt64Array)
* (UInt32Array, UInt32Array) --- `PartitionMode::Partitioned` for HashJoin 
* (UInt64Array, UInt64Array) --- Full join in NestedLoopJoin